File: aoc/utils/solve.py
import importlib
from typing import Callable

from rich import print

from aoc.utils.cli import Day, Language, Part, Year, parse_solve_cli_args
from aoc.utils.inputs import read_input_for_day

# TODO: https://github.com/marcelblijleven/adventofcode/blob/master/src/adventofcode/scripts/runner.py
# TODO: https://github.com/marcelblijleven/adventofcode/blob/master/src/adventofcode/scripts/benchmarks.py
# TODO: https://github.com/marcelblijleven/adventofcode/blob/master/src/adventofcode/scripts/generate_readme.py
# TODO: https://github.com/xavdid/advent-of-code-python-template/blob/main/advent

Input = list[str]
Answer = int | None
Example = tuple[Input, int]


def get_solution_function(year: Year, day: Day, part: Part, language: Language) -> Callable:
    try:
        module_name = f"aoc.{year}.{language}.{day}{'a' if part == 1 else 'b'}"
        solution_module = importlib.import_module(module_name)
        solution_function = getattr(solution_module, "solution")
        return solution_function
    except ModuleNotFoundError:
        raise ImportError(f"Module {module_name} not found.")
    except AttributeError:
        raise ImportError(f"Function 'solution' not found in module {module_name}.")


def calculate_answer(year: Year, day: Day, part: Part, language: Language) -> Answer:
    """TODO: handle solution module not found, function not found, etc."""
    solution = get_solution_function(year, day, part, language)
    input = read_input_for_day(year, day)
    # print(input)
    answer = solution(input)
    return answer


def print_answer(answer: Answer) -> None:
    if not isinstance(answer, int):
        print("Solution not yet implemented.")

    print(f"Answer: {answer}")


def submit_answer(answer: Answer) -> None:
    if not isinstance(answer, int):
        print("Solution not yet implemented. Submission cancelled.")
        print(f"Answer: {answer}")
        return

    print(f"Submitting answer: {answer}")
    print("TODO: Implement submit_answer")


def main() -> None:
    args = parse_solve_cli_args()
    answer = calculate_answer(args.year, args.day, args.part, "python")

    if args.submit:
        submit_answer(answer)
    else:
        print_answer(answer)


if __name__ == "__main__":
    main()

File: aoc/utils/typescript/solve.ts
import { type Day, parseSolveCliArgs, type Part, type Year } from './cli.ts'
import { readInputForDay } from './inputs.ts'

type Input = string[]
type Answer = number | null
type Example = [Input, number]

async function get_solution_function(year: Year, day: Day, part: Part): Promise<(x: string[]) => Answer> {
  const modulePath = `${Deno.cwd()}/aoc/${year}/typescript/${day}${part === 1 ? 'a' : 'b'}.ts`
  const module = await import(modulePath)
  return module.solution
}

async function get_answer(year: Year, day: Day, part: Part): Promise<Answer> {
  const solution = await get_solution_function(year, day, part)
  const input = readInputForDay(year, day)

  return solution(input)
}

function isInteger(answer: Answer): boolean {
  return typeof answer === 'number' && Number.isInteger(answer)
}

function printAnswer(answer: Answer): void {
  if (!isInteger(answer)) {
    console.log('🤔 Solution is not an integer. Have you solved this puzzle?')
  }

  console.log(`🔍 Answer: ${answer}`)
}

/**
 * See: https://docs.deno.com/api/deno/~/Deno.Command
 */
function submitAnswer(year: Year, day: Day, part: Part, answer: Answer) {
  if (!isInteger(answer)) {
    console.log('🤔 Solution is not an integer. Have you solved this puzzle?')
    console.log(`🔍 Answer: ${answer}`)
    return
  }

  console.log('Submitting answer...')

  const aocSessionCookieFile = `${Deno.cwd()}/.aoc-session-cookie`

  // see: https://github.com/scarvalhojr/aoc-cli?tab=readme-ov-file#usage-%EF%B8%8F
  const command = new Deno.Command(Deno.execPath(), {
    args: `aoc s -y ${year} -d ${day} -s ${aocSessionCookieFile} ${part} ${answer}`.split(' '),
  })

  command.outputSync()
  // const { code, stdout, stderr } = command.outputSync()
}

if (import.meta.main) {
  console.log('Running TypeScript solution')

  const args = parseSolveCliArgs()
  const answer = await get_answer(args.year, args.day, args.part)

  if (args.submit) {
    submitAnswer(args.year, args.day, args.part, answer)
  } else {
    printAnswer(answer)
  }
}
